aten/src/ATen/native/QuantizedLinear.cpp

```diff
@@ -110,51 +92,79 @@ Tensor fbgemm_linear_int8_weight_fp32_activation(
   int32_t weight_zero_point_int32 =
       static_cast<int32_t>(weight_zero_point.to<int64_t>());
 
-  // This is the end of the pipeline, pass the resulting matrix through
-  fbgemm::DoNothing<float, float> doNothingObj{};
-
   auto bias_contig = bias.contiguous();
 
-  // After the uint8 * int8 matrix multiplication is performed, this operation
-  // does:
-  //  1) Add in row and column offsets to the rows and columns, respectively
-  //  2) Dequantize the results into floating point
-  //  3) Add in the bias term
-  fbgemm::ReQuantizeForFloat</*FUSE_RELU*/false> outputProcObj(
-      /*nextop=*/doNothingObj,
-      /*Aq_scale=*/q_params.scale,
-      /*Bq_scale=*/&weight_scale_float,
-      /*Aq_zero_point=*/q_params.zero_point,
-      /*Bq_zero_point=*/&weight_zero_point_int32,
-      /*row_offsets=*/packA.getRowOffsetBuffer(),
-      /*col_offsets=*/col_offsets.data_ptr<int32_t>(),
-      /*bias=*/bias_contig.data_ptr<float>(),
-      /*nCol=*/N);
-
+  // The resulting matrix here is 2-D, let's view it with the original
+  // left hand dimensions of the input.
+  // Here are two examples:
+  // 1. If the input tensor is {M, K}, the output tensor is {M, N}.
+  // 2. If the input tensor is {b, M, K}, the output tensor is {b, M, N}.
+  std::vector<int64_t> out_sizes = input.sizes().vec();
+  out_sizes.back() = N;
   // Allocate output Tensor and a buffer for fbgemmPacked to use
   auto output = at::zeros(
```

Comment:
You can actually use empty and empty_like here, which should be more efficient.

Reply:
Thanks! Sorry I just saw this.

This is intentional...We don't want to have too much optimization efforts on this legacy op (although I saw @BIT-silence has put a PR to optimize that). It would be great if you can encourage the users to move to the new `torch.quantization.quantize_dynamic` API and optimize the new `quantized::linear_dynamic` op.

Attached the `empty` and `empty_like` optimizations for the ops in the new API:
https://github.com/pytorch/pytorch/pull/27002
https://github.com/pytorch/pytorch/pull/27005